ISPN-11784 Combine netty and non blocking thread pools

wburns opened the pull request:
* Move to only using IO event loop
* Change lifecycle to be controlled by component registry

https://issues.redhat.com/browse/ISPN-11784
https://issues.redhat.com/browse/ISPN-11883 (Do not close this one though)

wburns commented:
Resurrecting this now that 12.1.x has been branched and master is solely 13.

I am guessing some tests will fail the first few runs.

wburns commented:
Test failures seem to be unrelated. I will be running this a few more times and double checking to make sure there are no others though.

wburns commented:
run performance tests please

wburns commented:
run performance tests please
cs-hotrod
cs-rest

diegolovison commented:
run performance tests please
cs-hotrod
cs-rest

wburns commented:
The perf repo run looks good from the performance perspective.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/factories/NettyIOFactory.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @author Pedro Ruivo
+ * @author William Burns
+ * @since 11.0
```
```suggestion
 * @since 13.0
```

pruivo commented on server/rest/src/main/java/org/infinispan/rest/resources/CounterResource.java:
```diff
@@ -174,8 +192,15 @@ public Invocations getInvocations() {
+      NonRecursiveEventLoopGroup.reserveCurrentThread();
+      CounterConfiguration configuration;
+      try {
+         configuration = counterManager.getConfiguration(counterName);
```
you have `getConfigurationAsync(String)` if you want :) you would have to change this method :(

edit: same below :)

wburns replied:
That would be preferable, we don't want to use `NonRecursiveEventLoopGroup.reserveCurrentThread` if we don't need to. It is more of a bandaid until the underlying issue can be solved.

wburns replied:
Ah, I didn't see it because it wasn't on the interface :)

pruivo commented on server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java:
```diff
@@ -60,23 +61,33 @@ public Invocations getInvocations() {
-                  .addOptionalParameter("appenders", appenders)
-                  .subject(request.getSubject())
-      ).handle((o, t) -> handle(t));
+      NonRecursiveEventLoopGroup.reserveCurrentThread();
```
why do you need this for tasks? they are not blocking and this thread does not wait for the result.

wburns replied:
It isn't just about blocking vs non blocking. The problem is the combination of netty event loop picking a specific thread to run (which can include one that was previously used) on during submission and the fact that not all of our code is non blocking (the `runTask` invokes blocking code).

So what can happen is in this case non blocking thread 2 runs this code which calls `engine.handles(String)` which does a remote invocation that it blocks on. If the remote execution then calls back into this node and selects thread 2 to handle the request it deadlocks, since thread 2 is blocking waiting it can't run a new task. The `NonRecursiveEventLoopGroup.reserveCurrentThread` makes it so that the current thread can't be chosen by the netty event loop until it is released. This is just a workaround until the entire method can be made non blocking. Due to the deadlock nature, it can't be solved purely by not having a block hound exception.

pruivo replied:
`reserveCurrentThread` seems to "reserve" an `executor` and not a single thread. 
I guess if you invoke `reserveCurrentThread` recursively, eventually you are out of "executors".

wburns replied:
> `reserveCurrentThread` seems to "reserve" an `executor` and not a single thread.
With netty a given event loop is tied to a specific thread, so it is.

> I guess if you invoke `reserveCurrentThread` recursively, eventually you are out of "executors".
Yes, that is entirely possible to run out of event loops and thus threads if it was overused. That is the reason we want to use this very sparingly.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java:
```diff
@@ -0,0 +1,146 @@
+
+         return eventExecutor;
+      }
+      log.fatalf("No executor was found - replaying to trace issue - current reserveExecutors are %s", reserveExecutors);
```
I think the `fatal` messages have to be defined in the `Log` interface, right?

wburns replied:
I am torn, because this is will have been caused by a programming bug and the user can't really do anything about it. I can change them to DEBUG I guess.

pruivo replied:
the question is like: how does it help you find "a bug"?

wburns replied:
If you are asking why have the trace messages at all and just throw the exeception then the answer is it tells you why all the various executors were not able to be chosen, either by being the same event loop or it was reserved.

We could just remove the log messages I guess. This shouldn't really happen now, it was very helpful while finding the various areas. But since it should only show up for new code, it is probably fine.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java:
```diff
@@ -0,0 +1,146 @@
+            continue;
+         }
+
+         log.fatalf("Found a matched executor %s, possibly concurrent reservation?", eventExecutor);
```
can't you return the `eventExecutor` instead of throwing an `IllegalStateException`

wburns replied:
I don't want to. This was a sign that there was a bug, I would rather find it sooner rather than later.

pruivo replied:
> This was a sign that there was a bug

not really, it could be just bad luck

It looks like `next()` is not predictable (i.e. it is a shared counter, so requests from other threads will affect the return value in this loop) so you may never test all the executors in your loop. (worst case, it can return the same executor `loopAttempts` times)

wburns replied:
No, it is definitely a bug. If this is ever thrown that means it cannot use an executor. All the implementations of the netty event loop chooser by default will touch each event once in a loop of the same size. They use a modulus or power of two by incrementing a value. If they ever changed the implementation of the event executor chooser, then I agree it could be luck (which is one reason I loop through total number times two just in case). But as it is now it is very deterministic.

pruivo replied:
maybe I'm missing something... 
It looks like this code https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java#L73 is invoked when `next()` is invoked. 

If you have 2 threads submitting a task, one thread could get the executors in the even position and the other in the odd as an example. I don't see any code that suggests that all executors will be touched.

wburns replied:
I was referring to a unit test which would be hard pressed to run into this. But actually this reminds me why I had the replay logic with tracing so it would expose if this ever did happen :) Since the message you linked would be printed in such a case.

By the  way I forced it to two threads for runs in radargun and multiple times in CI and hasn't run across this yet :)

But this just reinforces using the reserve as little as possible.

wburns replied:
But thinking on it more I don't know if I am being critical enough for the concurrency case. Let me fix up the config ones at least and I will take a peak at the others more closely. The retrieving cache and counter cache ones should be extremely  rare and tasks currently are as well. But I will see if we can reduce this further.

wburns replied:
I can provide a custom event executor chooser which should remove the randomness and the only way to get the exception would be if all threads are reserved.

wburns replied:
Then the only way to run into the issue is having N concurrent calls to `reserveCurrentThread` where N is the number of event loops in the system.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java:
```diff
@@ -0,0 +1,146 @@
+      try {
+         shutdownGracefully().get(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+         log.debug("Interrupted while waiting for event loop group to shut down");
```
reset thread interrupt signal? `Thread.currentThread().interrupt();`?

wburns replied:
Sure.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java:
```diff
@@ -0,0 +1,146 @@
+      } catch (InterruptedException e) {
+         log.debug("Interrupted while waiting for event loop group to shut down");
+      } catch (ExecutionException e) {
+         throw new CacheException(e);
```
wrap the `cause` exception instead of `ExecutionException`?

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java:
```diff
@@ -0,0 +1,146 @@
+import io.netty.util.internal.ThreadExecutorMap;
+
+/**
+ * This event loop group prevents any tasks submitted via {@link java.util.concurrent.ExecutorService} API from
```
I guess `invokeAll` and `invokeAny` from `ExecutorService` are missing

wburns replied:
I didn't think of it because netty doesn't implement these methods in a thread pool way imo :( They just send all tasks to a single event loop. I would say we can just change the doc that only submit and execute are affected I guess.

wburns commented:
I have ran another run on CI to test for the issues with reserving the current thread and blocking in general. Will have to see what it finds tomorrow.

pruivo commented:
I've changed the `CounterResource.java` and removed the "reserve thread" stuff. Diff is here: https://gist.github.com/pruivo/44f171f6d256445750b20ac8ce02bd86

wburns commented on server/core/src/main/java/org/infinispan/server/core/factories/NettyEventLoopFactory.java:
```diff
@@ -0,0 +1,47 @@
+               globalConfiguration.transport().nodeName(), shortened(KnownComponentNames.NON_BLOCKING_EXECUTOR));
+      }
+
+      Executor executorToUse = new ThreadPerTaskExecutor(threadFactory);
```
Yeah, I was dabbling with replacing the event chooser, but sadly Netty uses terrible telescoping constructors and don't have all the various options I wanted :(

wburns commented:
> I've changed the `CounterResource.java` and removed the "reserve thread" stuff. Diff is here: https://gist.github.com/pruivo/44f171f6d256445750b20ac8ce02bd86

The reason I didn't do this originally iirc is because this adds quite a bit of latency for retrieving a cache, which the majority of the time will not block (only the first time it is retrieved). But looking more closely HotRodServer should be fine with this as it caches it after subsequent retrievals.

I am thinking for counters maybe I can just add a method that can return a null counter and not create it on request. Then if the counter isn't present it can invoke the blocking method in the blocking code, WDYT @pruivo ?

pruivo commented:
> I am thinking for counters maybe I can just add a method that can return a null counter and not create it on request. Then if the counter isn't present it can invoke the blocking method in the blocking code, WDYT @pruivo ?

sounds good to me.

wburns commented:
@pruivo I have reworked it to no longer have the reserve thread by not blocking in those spots. I have left the cache retrieval as is, as a normal server will not run into this issue of blocking. Tests run into it because they are manually invoking `createConfiguration` which is not done on a server.

wburns commented:
Test failures all appear unrelated.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/backup/BackupWriter.java:
```diff
@@ -102,28 +102,32 @@ void cleanup() {
+                  resources.forEach(r -> r.writeToManifest(manifest));
+                  storeProperties(manifest, "Container Properties", containerRoot.resolve(CONTAINERS_PROPERTIES_FILE));
+               },
+               "create-manifest"
```
same here?

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NettyTransport.java:
```diff
@@ -71,6 +75,7 @@
 
    //-1 if not set
    private int nettyPort = -1;
+   private boolean shouldShutdownIOGroup;
```
nit:
the value of `shouldShutdownIOGroup` is always `cacheManager == null`. IMO, the field is no required.

wburns replied:
Let me take a look again. There was something with the single port router where this is null, let me check again.

wburns replied:
oh gotcha, you mean just check cacheManager, sure.

pruivo commented on server/core/src/main/java/org/infinispan/server/core/transport/NettyTransport.java:
```diff
@@ -340,10 +354,18 @@ public void updateTotalBytesRead(int bytes) {
+      return eventLoop;
+   }
+
+   public static MultithreadEventLoopGroup buildEventLoop(int nThreads, Executor executor, String configuration) {
```
I think this method is not used anymore.

wburns replied:
Thought I had removed that, let me do that

pruivo commented on tasks/scripting/src/main/java/org/infinispan/scripting/impl/ScriptingTaskEngine.java:
```diff
@@ -39,6 +40,12 @@ public String getName() {
+   @Override
+   public CompletionStage<List<Task>> getTasksAsync() {
+      BlockingManager blockingManager = scriptingManager.cacheManager.getGlobalComponentRegistry().getComponent(BlockingManager.class);
+      return blockingManager.supplyBlocking(this::getTasks, "ScriptingTaskEngine - getTasksAsync");
```
isn't `Cache.stream()` non blocking? could it be used instead of sending to a new thread?

wburns replied:
The call to `stream` is, but any subsequent method invocations are not in this case `forEach` is not. That is what the ClusterPublisherManager is for, but we don't have a nice API for it yet, so I figured this was simpler for now.

pruivo reviewed (commented):
Just some minor comments :)

wburns commented:
Updated.

pruivo commented:
it looks like `org.infinispan.client.hotrod.ExecTest` failure is related.

wburns commented:
> it looks like `org.infinispan.client.hotrod.ExecTest` failure is related.

Yeah, it appears it is related to script manager invoking scripts on the non blocking thread and the script blocking. I will change it so scripts are invoked on a blocking thread instead. Also I will remove the blockhound exclusion for script manager.

pruivo commented:
I don't see this change: `I have pushed a new commit that changes script manager so it only invokes a script on a blocking thread,`

wburns commented:
> I don't see this change: `I have pushed a new commit that changes script manager so it only invokes a script on a blocking thread,`

Ooops cherry picked the wrong commit, fixing :)

wburns commented:
CI is good now. However, it has exposed a new blocking code path.

```
      java.lang.AssertionError: Blocking call! jdk.internal.misc.Unsafe#park on thread Thread[HotRodFilterEventsTest-Client-88-1,5,main]
	at org.infinispan.util.CoreTestBlockHoundIntegration.lambda$applyTo$0(CoreTestBlockHoundIntegration.java:52)
	at reactor.blockhound.BlockHound$Builder.lambda$install$8(BlockHound.java:383)
	at reactor.blockhound.BlockHoundRuntime.checkBlocking(BlockHoundRuntime.java:89)
	at java.base/jdk.internal.misc.Unsafe.park(Unsafe.java)
	at java.base/java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:885)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:917)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1240)
	at java.base/java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:267)
	at java.base/java.util.concurrent.ArrayBlockingQueue.offer(ArrayBlockingQueue.java:341)
	at java.base/java.util.AbstractQueue.add(AbstractQueue.java:95)
	at java.base/java.util.concurrent.ArrayBlockingQueue.add(ArrayBlockingQueue.java:326)
	at org.infinispan.server.hotrod.event.EventLogListener.onCreated(EventLogListener.java:72)
	at org.infinispan.server.hotrod.test.ClientHandler.channelRead(HotRodClient.java:1160)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Fixing this and will do a couple more runs again.

wburns commented:
Test runs in the past few runs all seem unrelated. I have already fixed up the commits, so this should be ready.

pruivo commented:
integrated! thanks @wburns !
